### ISSUE TITLE
enhancing the performance of very large installation of cacti through various main pollers and single url to access all instances of cacti

### ISSUE THREAD

**Issue by gj00354347** [Cacti/cacti#5396]
# Feature Request

Hi Team ,

we are monitoring different devices of Data Center like Switches , Routers , Load Balancers , Firewalls etc whose total number is around 4K and they are having around 0.5 Million graphs and around 1 Million Data Sources . SO cacti  allows at most  100 graphs to be viewed in a single page . we were thing of automating the L1 Monitoring of Cacti through Bot , where Bot is supposed to view each and every graph and need to report to INC tool if it finds and issue there.
But page loading time for each page is around 15-20 seconds and then  we have 5K pages for 0.5 Million graphs , so the Bot will have to traverse those 5K graphs which will take around 20 * 5K seconds and this will also put a huge load on cacti's performance .

so what we were thinking of is that we distribute the load over several Main pollers , like as of now we have only one main Poller and 9 remote pollers . the proposed solution will have 3-4 Main poller and the 9 Remote pollers will be associated with these Main Pollers .
the configuration will be like
RM1 & RM 2 --> MP1
RM3 -->MP2
RM4 & RM5 --> MP3
RM6, RM7 & RM8 --> MP4
RM9 -> MP5

RM= Remote Poller and MP=Main Pollers 

there will be a super url named https://superurl.com/cacti which will be same for all customers and that will access different instances of cacti named httpd://mainpoller1.com/cacti or https://mainpoller2.com/cacti or https://mainpoller3.com/cacti etc based on which device graphs are requisitioned .

in this way we will not only be able to make cacti fast but also be able to scale to provision many new devices in cacti .

the different version of components in use are 
mysql 10.5 
php 7.x 
httpd 2.4.34
OS - RHEL 7.9 Maipo
cacti 1.2.18
RAM - on Main poller around 100GB and on Remote poller 20-45 GB based on number of devices being polled from that Remote poller

## Is your feature request related to a problem? Please describe

A clear and concise description of what the problem is. Ex. I'm always
frustrated when [...]

## Describe alternatives you've considered

A clear and concise description of any alternative solutions or features you've
considered.

**Comment by netniV** [comment]
You can do this kind of thing if you implement a load balancer such as HAProxy and have rules based on authentication / state that decide which backend server will handle the front end request. 

Eg, without auth, set the backend to be bk_auth_server. With auth, set the backend to be bk_client_XXX where XXX is decided somehow based upon what the auth has provided, such as the sub path (/cacti1 or /cacti2). This would mean that although they appear on the same server, they are actually their own instances. 

I could see how that would allow you to scale and have small instances of cacti handling each client. One thing it won't do is allow you as an admin visibility of everyone at once. 

However, this kind of setup is unfortunately beyond the scope of the cacti project itself and would be something you'd have to handle.

**Comment by TheWitness** [comment]
@gj00354347,

We have a netscaler front-ending our cacti, and the database is separate.  3 web servers with each 88 threads and a bunch of memory all tied to a 3 way replicated gluster cluster.

We use RedHat HA to keep it stood up.  If you are planning on using a robot, it should not need a UI and likely wants to look at numbers, but heck, maybe an image is more important.  So, I would do it with a plugin.

I broached the idea of multiple whole Cacti's and a Super Cacti when I designed the multipoller interface in 1.x and we refined it in I.2.x.  I think it was a good compromise. I wanted to keep the RRDfile in a central place by design, though we were kind of backed into it by multiple confounding factors.  Modern storage has no issues with it.
